Added a Builder for ProxyHandler

CodingFabian opened the pull request:
Because @trustin and @normanmaurer agreed in #4125 that having a Factory for proxy handler is nice to have, here my first stab at it.

I marked this as WIP, because I want to have feedback on the following first, before I complete it with doc, test and commit message.
* I chose a builder over a factory, however it gracefully accepts empty values for username and password. The rationale is that this Builder is supposed to free users from checking all parameters and decide which class to construct, so validating them would turn this builder useless
* I used InetSocketAddress.createUnresolved - I assume that is common pattern in Netty, because resolving the address happens on usage not on construction, right?
* I am throwing RuntimeExceptions on unsupported proxy types. I checked a couple of other places, and this pattern seems ok, can you suggest a different exception?
* I am using an enum, which allows preventing getting an exception for type. I have not found other places in netty with a similar pattern. is that ok? Note that I have a lenient parser to also accept the string "socks5". Should the enum be a top level type?

Sorry for that many questions, but I want my contribution to fit netty style :)

normanmaurer commented on handler-proxy/src/main/java/io/netty/handler/proxy/ProxyHandlerBuilder.java:
```diff
@@ -0,0 +1,107 @@
+public final class ProxyHandlerBuilder {
+
+  public static enum ProxyType {
+    HTTP, SOCKS4, SOCKS5
```
Use camel-case

trustin replied:
I thought we are using UPPERCASED_NAME for enums?

normanmaurer replied:
huh... didn't we just change to camel-case in the ipfilter ?

trustin replied:
Huh? I thought that was a mistake in the old code and my comment was to
rename them to ACCEPT and DENY, and you issued a pr for the rename?

Scottmitch replied:
https://github.com/netty/netty/pull/4112#discussion_r37606073 .. I thought there was a mis queue on that PR... https://github.com/netty/netty/pull/4111

CodingFabian replied:
i am confused, could you give me a clear signal for the one or the other? thanks :)

normanmaurer commented on handler-proxy/src/main/java/io/netty/handler/proxy/ProxyHandlerBuilder.java:
```diff
@@ -0,0 +1,107 @@
+        return new ProxyHandlerBuilder(type);
+      }
+    }
+    throw new RuntimeException("Proxy type " + proxyType + " not supported.");
```
new IllegalArgumentException

CodingFabian replied:
done

normanmaurer commented on handler-proxy/src/main/java/io/netty/handler/proxy/ProxyHandlerBuilder.java:
```diff
@@ -0,0 +1,107 @@
+
+  public ProxyHandlerBuilder proxyAddress(SocketAddress proxyAddress) {
+    checkNotNull(proxyAddress, "proxyAddress");
+    this.proxyAddress = proxyAddress;
```
merge these two lines

CodingFabian replied:
done

normanmaurer commented on handler-proxy/src/main/java/io/netty/handler/proxy/ProxyHandlerBuilder.java:
```diff
@@ -0,0 +1,107 @@
+
+  public ProxyHandlerBuilder proxyAddress(String host, int port) {
+    checkNotNull(host, "proxyHost");
+    this.proxyAddress = InetSocketAddress.createUnresolved(host, port);
```
merge these two lines

CodingFabian replied:
done

normanmaurer commented on handler-proxy/src/main/java/io/netty/handler/proxy/ProxyHandlerBuilder.java:
```diff
@@ -0,0 +1,107 @@
+  }
+
+  public ProxyHandler build() {
+    checkNotNull(proxyAddress, "proxyAddress");
```
maybe throw an IllegalStateException would be better.

CodingFabian replied:
done

normanmaurer commented on handler-proxy/src/main/java/io/netty/handler/proxy/ProxyHandlerBuilder.java:
```diff
@@ -0,0 +1,107 @@
+        return new Socks5ProxyHandler(proxyAddress, username, password);
+      }
+    default:
+      throw new RuntimeException("Proxy type " + type + " not supported.");
```
new Error(..)

CodingFabian replied:
you really use java.lang.Error for that? While it is a problem which is similar to noclassdeferror, i find that pretty hard. done anyway :)

normanmaurer commented on handler-proxy/src/main/java/io/netty/handler/proxy/ProxyHandlerBuilder.java:
```diff
@@ -0,0 +1,107 @@
+      throw new RuntimeException("Proxy type " + type + " not supported.");
+    }
+  }
+
```
remove empty line

CodingFabian replied:
done

normanmaurer commented on handler-proxy/src/main/java/io/netty/handler/proxy/ProxyHandlerBuilder.java:
```diff
@@ -0,0 +1,107 @@
+ */
+public final class ProxyHandlerBuilder {
+
+  public static enum ProxyType {
```
use 4 spaces...

CodingFabian replied:
oops yeah, its  the wrong formatter :)

CodingFabian commented:
besides the minor code stuff, is the general pattern ok? if yes i can add test and docs.

bk1te commented:
What about proxy chaining ? Like in trustin's tests for proxy handler?

CodingFabian commented:
@bk1te would you mind explaining your suggestion? The Builder creates ProxyHandlers based on configuration. It doesnt add them to any handler chains or such.

bk1te commented:
If you create proxy handler via builder? Why don't create ProxyChainBuilder to create not just one handler. It is more general way for dealing with proxy. But it would be difficult to add in it http+ssl proxy

CodingFabian commented:
I have addressed all comments so far, added a test case which covers all lines except the enum Error case in build().
As far as I see you do not have strict javadoc rules, so I added where I usually add them (where it makes sense :-)).
Let me know if thats ok or needs fixups.
